### PR TITLE
realise mongodb interface

### PR DESCRIPTION
**Pull request by jianqin123**
add function on mongondb includes
 insert :insert data
update:update data
select :select data
delete :delete data
use the same interface with creatClient and Connection

### CODE REVIEW

**Comment by jianqin123** [comment]
I am new to this area

File: test/mongodb.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        } )
+   })
+    
+ 
+
+  
+
+   it("update MongoDB",function(done){
+           var where ={"name23":"www..com"} ;
+           var update={$set:{"nadfd":"fuck"}}
```
Comment (yinrong) [inline test/mongodb.js]: remove

**Comment by yinrong** [comment]
commit message is bad

**Comment by jianqin123** [comment]
thanks  for review ,I will try my best to fix it on time .

**Comment by jianqin123** [comment]
Optimized the code

File: lib/define/MongoDB.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
+var singleton_pools = {}
+module.exports.Context = function (config) {
+    assert.equal(config.type, 'MongoDB')
```
Comment (yinrong) [inline lib/define/MongoDB.js]: please read wiki. two spaces here

Comment (yinrong) [inline lib/define/MongoDB.js]: I don't see README here. This is mentioned again.

File: lib/define/MongoDB.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    const pool = singleton_pools[config_key]
+    this.getConnection = function () {
+        log.info('getConnection MongoDB')
+        var ret = new MongoConnection(pool, config.readonly)
+        return ret
+    }
+}
+
+function MongoConnection(pool, readonly) {
+    this._pool = pool.DB_CONN_STR
```
Comment (yinrong) [inline lib/define/MongoDB.js]: please read wiki. no abbreviation here. Also, this config is duplicated

File: lib/define/MongoDB.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+}
+/**
+ * select all data from set
+ * @param set :colleciton
+ * @return ：return -1 if error，otherwise return all data 
+ */
+MongoConnection.prototype.selectAll = function (set, callback) {
+    var server = new mongodb.Server(this._host, this._port, { auto_reconnect: true });
+    var db = new mongodb.Db(this._db, server, { safe: true });
```
Comment (yinrong) [inline lib/define/MongoDB.js]: remove all colons

File: lib/define/MongoDB.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+ * @param return -1 if error ,otherwise return data
+ */
+MongoConnection.prototype.select = function (set, where, callback) {
+    if (this._readonly) {
+        return errNotAllowed(callback)
+    }
+    var server = new mongodb.Server(this._host, this._port, { auto_reconnect: true });
+    var db = new mongodb.Db(this._db, server, { safe: true });
+
+    db.open(function (err, db) {
```
Comment (yinrong) [inline lib/define/MongoDB.js]: there should be NO duplication of source code. `server ... db.open(...)` is duplicated multiple times

File: lib/define/MongoDB.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        }
+    });
+    mongodb.MongoClient.connect(this._pool, numberOfRetries = retryNum, callback)
+}
+
+const Q = require('q')
+const retryNum = 1
+const child_process = require('child_process')
+const _ = require('lodash')
+const bson = require('bson');
```
Comment (yinrong) [inline lib/define/MongoDB.js]: ?

File: lib/define/MongoDB.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            db.close();
+            return "insert into  success"
+        }
+    });
+    mongodb.MongoClient.connect(this._pool, numberOfRetries = retryNum, callback)
+}
+
+const Q = require('q')
+const retryNum = 1
+const child_process = require('child_process')
```
Comment (yinrong) [inline lib/define/MongoDB.js]: ?

File: lib/define/user.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * select all data from set
+ * @param collection
+ * @return return -1 when fail ,otherwise return all data
+ */
+function findCollectionNoCondition(collection) {
+    console.log("come here for open collection")
+    var deferred = Q.defer();
+    collection.find().toArray(function (errfind, cols) {
+        if (!errfind) {
```
Comment (yinrong) [inline lib/define/user.js]: please read wiki carefully.

File: lib/define/MongoDB.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            user.insertCollection(collection, data, callback)
+            db.close();
+            return "insert into  success"
+        }
+    });
+    mongodb.MongoClient.connect(this._pool, numberOfRetries = retryNum, callback)
+}
+
+const Q = require('q')
+const retryNum = 1
```
Comment (yinrong) [inline lib/define/MongoDB.js]: violate naming convention

File: lib/define/MongoDB.js [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+const child_process = require('child_process')
+const _ = require('lodash')
+const bson = require('bson');
+const mongodb = require('mongodb')
+const QUERY_READONLY_ALLOW = ['SELECT']
+const util = require('util')
+const mysql = require('mysql')
```
Comment (yinrong) [inline lib/define/MongoDB.js]: ?

File: lib/define/MongoDB.js [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+const crypto = require('crypto')
+const async = require('async')
+const assert = require('assert')
+const EventEmitter = require('events')
+const debug = require('debug')('sqlx')
+var local = {}
+function __log() {
+    debug(util.format.apply(null, arguments))
+}
+const log = {
```
Comment (yinrong) [inline lib/define/MongoDB.js]: do not copy code

File: lib/define/MongoDB.js [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+const Q = require('q')
+const retryNum = 1
+const child_process = require('child_process')
+const _ = require('lodash')
+const bson = require('bson');
+const mongodb = require('mongodb')
+const QUERY_READONLY_ALLOW = ['SELECT']
+const util = require('util')
```
Comment (yinrong) [inline lib/define/MongoDB.js]: already defined in $, do not require again

File: lib/define/user.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+ * insert data
+ * @param collection
+ * @param tmp data to insert
+ * @return return -1 when fail ,else return collection
+ */
+function insertCollection(collection, tmp,callback) {     
+    collection.insert(tmp, { safe: true }, function (err, result) {
+        if (err) {
+            console.log('fail to set' + tmp);
+            // return -1;
```
Comment (yinrong) [inline lib/define/user.js]: ?

do not comment unless it's necessary

File: lib/define/MongoDB.js (hunk 2 above)
Comment (yinrong) [inline lib/define/MongoDB.js]: MongoDB violates our naming convention

File: lib/define/MongoDB.js [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+ * update data satisfied from old value to new value
+ * @param set :collection
+ * @param where:old value
+ * @param update :new value 
+ *
+ */
+
+MongoConnection.prototype.update = function (set, where, update, callback) {
+    if (this._readonly) {
+        return -1;
```
Comment (yinrong) [inline lib/define/MongoDB.js]: why return?

Comment (yinrong) [inline lib/define/MongoDB.js]: no returns for any callback-style functions

File: lib/define/MongoDB.js (hunk 5 above)
Comment (yinrong) [inline lib/define/MongoDB.js]: put shared code in common function or files.

File: lib/define/MongoDB.js [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+  }
+  const connect_info = singleton_pools[config_key]
+  this.getConnection = function () {
+    log.info('getConnection MongoDB')
+    var ret = new MongoConnection(connect_info, config.readonly)
+    return ret
+  }
+}
+function MongoConnection(connect_info, readonly) {
+  this._addreess = connect_info.address
```
Comment (yinrong) [inline lib/define/MongoDB.js]: _addreess

File: lib/define/MongoDB.js [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+  this._host = connect_info.host
+  this._port = connect_info.port
+  this._db = connect_info.db;
```
Comment (yinrong) [inline lib/define/MongoDB.js]: ;

File: lib/define/MongoDB.js [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+}
+var user = require('./user')
+/**
+ * select data satified the condition
+ * @param set collection
+ * @param where condition * 
+ */
+var connectToDatabase=function(host,port,database){
+  var server = new mongodb.Server(host, port, { auto_reconnect: true });
```
Comment (yinrong) [inline lib/define/MongoDB.js]: ;

File: lib/define/MongoDB.js [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+  var db = new mongodb.Db(database, server, { safe: true });
+  return db
+}
+MongoConnection.prototype.select = function (set, where, callback) {
+  if (this._readonly) {
+    console.log("only read is allowed")
+  } 
+  var db=connectToDatabase(this._host,this._port,this._db);
+  db.open(function (err, db) {
```
Comment (yinrong) [inline lib/define/MongoDB.js]: put db.open in a common function

File: lib/define/MongoDB.js [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+var connectToDatabase=function(host,port,database){
+  var server = new mongodb.Server(host, port, { auto_reconnect: true });
+  var db = new mongodb.Db(database, server, { safe: true });
+  return db
+}
+MongoConnection.prototype.select = function (set, where, callback) {
+  if (this._readonly) {
+    console.log("only read is allowed")
+  } 
+  var db=connectToDatabase(this._host,this._port,this._db);
```
Comment (yinrong) [inline lib/define/MongoDB.js]: space?

**Comment by jianqin123** [comment]
ok,I will cherish the last chance.

**Pull request closed without merging**